doc/source/data/iterating-over-data.rst

```diff
@@ -0,0 +1,273 @@
+        :sync: NumPy
+
+        .. testcode::
+
+            import ray
+
+            ds = ray.data.read_images("example://image-datasets/simple")
+
+            for batch in ds.iter_batches(batch_size=2, batch_format="numpy"):
+                print(batch)
+
+        .. testoutput::
+
+            {'image': array([[[[...]]]], dtype=uint8)}
+            ...
+            {'image': array([[[[...]]]], dtype=uint8)}
+
+    .. tab-item:: pandas
+        :sync: pandas
+
+        .. testcode::
+
+            import ray
+
+            ds = ray.data.read_csv("s3://anonymous@air-example-data/iris.csv")
+
+            for batch in ds.iter_batches(batch_size=2, batch_format="pandas"):
+                print(batch)
+
+        .. testoutput::
+            :options:+NORMALIZE_WHITESPACE
+
+               sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  target
+            0                5.1               3.5                1.4               0.2       0
+            1                4.9               3.0                1.4               0.2       0
+            ...
+               sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  target
+            0                5.1               3.5                1.4               0.2       0
+            1                4.9               3.0                1.4               0.2       0
+
+    .. tab-item:: Torch
+        :sync: Torch
+
+        .. testcode::
+
+            import ray
+
+            ds = ray.data.read_images("example://image-datasets/simple")
+
+            for batch in ds.iter_torch_batches(batch_size=2):
+                print(batch)
+
+        .. testoutput::
+
+            {'image': tensor([[[[...]]]], dtype=torch.uint8)}
+            ...
+            {'image': tensor([[[[...]]]], dtype=torch.uint8)}
+
+    .. tab-item:: TensorFlow
+        :sync: TensorFlow
+
+        .. testcode::
+
+            import ray
+
+            ds = ray.data.read_csv("s3://anonymous@air-example-data/iris.csv")
+
+            tf_dataset = ds.to_tf(
+                feature_columns="sepal length (cm)",
+                label_columns="target",
+                batch_size=2
+            )
+            for features, labels in tf_dataset:
+                print(features, labels)
+
+        .. testoutput:
+
+            tf.Tensor([5.1 4.9], shape=(2,), dtype=float64) tf.Tensor([0 0], shape=(2,), dtype=int64)
+            ...
+            tf.Tensor([5.1 4.9], shape=(2,), dtype=float64) tf.Tensor([0 0], shape=(2,), dtype=int64)
+
+For more information on working with batches, see
+:ref:`Transforming batches <transforming-batches>` and
+:ref:`Inspecting batches <inspecting-batches>`.
+
+.. _iterating-over-batches-with-shuffling:
+
+Iterating over batches with shuffling
+=====================================
+
+:class:`Dataset.random_shuffle <ray.data.Dataset.random_shuffle>` is slow because it
+shuffles all rows. For better performance, shuffle a subset of rows while iterating.
```

Comment:
when would I want to use `random_shuffle` then?

I think we should clarify this more "If a full global shuffle is not required, you can shuffle a subset of rows up to a provided buffer size during iteration by specifying `local_shuffle_buffer_size`. While this is not a true global shuffle like `random_shuffle`, it is more performant as it does not require excessive data movement"